server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -315,4 +378,96 @@ protected boolean isFresh(Long currentKey, Long newKey) {
             return newKey <= currentKey;
         }
     }
+
+    private static boolean doRemotes(ClusterStatsRequest request) {
+        return CCS_TELEMETRY_FEATURE_FLAG.isEnabled() && request.doRemotes();
+    }
+
+    private class ActionListenerWithRemotes implements ActionListener<ClusterStatsResponse> {
+        private final ActionListener<ClusterStatsResponse> listener;
+        private final ActionFuture<Map<String, RemoteClusterStatsResponse>> remoteFuture;
+
+        ActionListenerWithRemotes(ActionListener<ClusterStatsResponse> listener, ClusterStatsRequest request) {
+            this.listener = listener;
+            remoteFuture = getStatsFromRemotes(request);
+        }
+
+        private Map<String, RemoteClusterStatsResponse> resolveRemoteClusterStats() {
+            try {
+                return remoteFuture.actionGet();
```

Comment:
This blocks a thread while waiting for the remote responses, we definitely don't want that. Please make this whole thing nonblockingly async.